NF: introduce TaskWithoutCol to simplify code

Arthur-Milchior opened the pull request:
Checking whether it's one specific class do not seems very elegant.
Ideally, having a method `boolean RequiresTheCollection()` would be ideal. However, I like the simplicity of having a
single method, as it will allow to use lambda in future commit. This keeps the code simple and still cleaner

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -1262,7 +1267,7 @@ public RenderBrowserQA(CardBrowser.CardCollection<CardBrowser.CardCache> cards,
 
 
-    public static class RepairCollectionn extends Task<Void, Boolean> {
+    public static class RepairCollectionn extends TaskWithoutCol<Void, Boolean> {
```
nit (not for this PR): should be `RepairCollection`

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -105,6 +105,11 @@
+    /**
+     * Similar to task, but indicate that it should be executed even if the collection is not opened.
+     */
+    public abstract static class TaskWithoutCol<ProgressBackground, ResultBackground> extends Task <ProgressBackground, ResultBackground>{}
```
It's unintuitive that `TaskWithoutCol` calls `getCol()`, we should have a method of `getCol()` which doesn't log if it fails.

We should have these accept an `Optional<Collection>` in the `task` method

Arthur-Milchior replied:
Actually, one of the task begin by closing the collection if its opened. The other one use the collection to get context, and that's it. Probably not really necessary in the first place then.

I don't see the point of using Optional<Collection> since we always pass a collection

david-allison-1 replied:
Because it signals to the developer that the value is optional: either that or a `@Nullable`

david-allison-1 reviewed (commented):
This raises a bug for a separate issue:

If `getCol()` returns null, this crashes

https://github.com/ankidroid/Anki-Android/blob/143d7bbf7c9aa87f842d44c240c07b7cb2bd9285/AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java#L1374-L1377

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -105,6 +105,11 @@
+    /**
+     * Similar to task, but indicate that it should be executed even if the collection is not opened.
+     */
+    public abstract static class TaskWithoutCol<ProgressBackground, ResultBackground> extends Task <ProgressBackground, ResultBackground>{}
```
Actually, in one case, it needs the collection if it exists to close it. 

The difference is only that those task can proceed if the collection does not exists while all other tasks are aborted if they can't access the collection

david-allison-1 replied:
`WithoutCol` doesn't make sense to me if the method can take a `col`, is there a better name that we can use?

Arthur-Milchior commented:
I changed it. I added a method. It's cleaner and do not rely on class at all anymore.

I still hope to transform some tasks into lambda, but with default method it won't actually be a trouble

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:
```diff
@@ -103,6 +103,10 @@
     public abstract static class Task<ProgressBackground, ResultBackground> {
         protected abstract ResultBackground task(@NonNull Collection col, @NonNull ProgressSenderAndCancelListener<ProgressBackground> collectionTask);
+
+        protected boolean requireAnOpenCollection() {
```
`requiresOpenCollection`

Arthur-Milchior replied:
Done

david-allison-1 reviewed (approved):
LGTM
